### PR TITLE
SNOW-607466: Fix the limited length of string column in `create_dataframe`

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

**Review by sfc-gh-sfan — APPROVED** [review]

File: tests/integ/test_dataframe.py [hunk 1]
```diff
@@ -1583,3 +1583,18 @@ def test_unpivot(session, column_list):
...
+def test_create_dataframe_string_length(session):
+    table_name = Utils.random_name_for_temp_object(TempObjectType.TABLE)
+    df = session.create_dataframe(["ab", "abc", "abcd"], schema=["a"])
+    df.write.mode("overwrite").save_as_table(table_name, create_temp_table=True)
+    datatype = json.loads(
+        session.sql(f"show columns in {table_name}").collect()[0]["data_type"]
+    )
+    assert datatype["type"] == "TEXT"
+    assert datatype["length"] == 2**20 * 16  # max length (16 MB)
+    session.sql(f"insert into {table_name} values('abcde')").collect()
```
Comment (sfc-gh-jdu) [inline tests/integ/test_dataframe.py:1597]: Unrelated to this PR: I found we have `update` and `delete` method for table https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/_autosummary/snowflake.snowpark.html#snowflake.snowpark.Table, but we don't have `insert` method (we have `insert` when merging two tables), so I have to use SQL here. Do you think we should add it? Note that scala snowpark and databricks' delta table doesn't have it. https://docs.databricks.com/delta/delta-update.html#language-python

**Comment by sfc-gh-jdu** [comment]
@sfc-gh-bli FYI the simple fix works

File: src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py [hunk 2]
```diff
@@ -200,5 +204,5 @@ def to_sql_without_cast(value: Any, data_type: DataType) -> str:
     if value is None:
         return "NULL"
     if isinstance(data_type, StringType):
-        return f"""{value}"""
+        return f"'{value}'"
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py:207]: This is a bug that we didn't realize before https://github.com/snowflakedb/thundersnow/blob/535a433f3157de6a43eea600797ab5c50cca5426/src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala#L115

**Review by sfc-gh-yixie — APPROVED** [review]

File: src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py (hunk 2 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py:207]: Nope, it's possible that `value` here is numeric, but `data_type` is `StringType`.

File: src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py [hunk 3]
```diff
@@ -87,7 +87,11 @@ def to_sql(value: Any, datatype: DataType) -> str:
 
     # Not nulls
     if isinstance(value, str) and isinstance(datatype, StringType):
-        return str_to_sql(value)
+        return (
+            f"{str_to_sql(value)} :: string"
```
[the remark this answers is not in the document]
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py:94]: added

**Pull request merged**